### ISSUE TITLE
linting on fly

### ISSUE THREAD

**Issue by sstaub** [hebaishi/linter-gcc#21]
A linting on fly (while typing instead after saving a file) is really missing.

**Comment by hebaishi** [comment]
Linting on-the-fly has been deliberately disabled since linting is achieved by compilation, which can be both time and cpu-intensive. Compiling a source file with every keystroke is therefore, not feasible.

**Comment by sstaub** [comment]
Understand, it is possible to start a lint process after finishing a line with Enter/Return

**Comment by hebaishi** [comment]
Unfortunately, the linter api (which linter-gcc relies on) doesn't give you this level of flexibility. It only allows you to turn linting on-the-fly on and off. Turning it on means the source file is recompiled with every keystroke, which is not feasible.

**Comment by sstaub** [comment]
Ok, how can I turn the option on?

**Comment by hebaishi** [comment]
The linter doesn't work when you turn on linting on-the-fly. You simply get no errors even when are errors in your source file. If you still wish to turn the option on, you will need to fork linter-gcc and turn make the change yourself.

**Comment by sstaub** [comment]
I found the option lintOnFly, is this the only change i need to make?

Thank you for your help

sstaub

**Comment by steelbrain** [comment]
@hebaishi We actually have several types of linter providers. There are traditional ones that are triggered on keypress or save, and there are indie ones that are completely managed by the provider. The provider can set or remove messages at any time.

See Linter-Indie API in [Linter docs](https://github.com/atom-community/linter/wiki/Linter-Indie-API)

**Comment by hebaishi** [comment]
Hmm... this is very interesting. Ok I will have a play and try to implement a lint on-the-fly feature that works when the user presses Enter.

**Comment by hebaishi** [comment]
Hello @sstaub. I am working on an update to linter-gcc that will have linting on save switched on by default, but allow the user to turn on linting on-the-fly as well.

**Comment by sstaub** [comment]
Thank you very much, that will us help. I am working with plattformIO and this is missing killer feature. Because of working with embedded systems like arduino and mbed, resources are not such a big problem.

sstaub

**Comment by hebaishi** [comment]
I have got it working pretty well now. The only thing to note is that it saves the file after every change. The automatic saving only happens if linting on-the-fly is enabled, but getting it to work another way would involve either creating a temporary file somewhere, and making sure that the include paths work properly, or somehow trying to pipe to g++ from stdin, which would be messy. What do you think?

**Comment by steelbrain** [comment]
@hebaishi We also have a `tempFile` helper in [`atom-linter`](https://github.com/AtomLinter/atom-linter#api)

**Comment by hebaishi** [comment]
@steelbrain you guys really have thought of everything! Ok back to atom!

**Comment by hebaishi** [comment]
Linting on-the-fly is on the way! Writing the tests at the moment, update will likely come tomorrow.

**Comment by hebaishi** [comment]
@steelbrain I'm having an issue with linting on-the-fly. The linter works as expected - when a C/C++ file is opened and edited, the error messages are updated correctly. I have added a cleanup function that removes error messages from closed TextEditors, and this seems to be the issue. Saving the file and closing it works as expected - the messages of the closed file are removed. But if you open a file, edit it, and close it without saving, you get this error in the console:
```
Uncaught (in promise) Error: Given editor is not really an editor
    at MessageRegistry.set (/home/husam/.atom/packages/linter/lib/message-registry.js:41:15)
    at /home/husam/.atom/packages/linter/lib/linter.coffee:37:17
    at Function.module.exports.Emitter.simpleDispatch (/usr/share/atom/resources/app.asar/node_modules/event-kit/lib/emitter.js:25:14)
    at Emitter.module.exports.Emitter.emit (/usr/share/atom/resources/app.asar/node_modules/event-kit/lib/emitter.js:125:28)
    at /home/husam/.atom/packages/linter/lib/linter-registry.js:66:26
```
I have pushed my changes to a new branch called 'test'. Do you think you could help me with this?

**Comment by hebaishi** [comment]
@steelbrain I think I have idenfied the problem. The issue is that I had atom's auto-save switched on. It somehow interacts with linter and causes this problem when a destroyed TextEditor is auto-saved. When I switch it off, everything works perfectly

**Comment by hebaishi** [comment]
Ok. Package updated with linting on-the-fly. Please update to start using it!

**Comment by sstaub** [comment]
Cool, thank you very much, this makes our work much easier.
A restart of atom is necessary.

sstaub
